src/main/java/com/pinterest/secor/uploader/GsUploadManager.java

```diff
@@ -111,11 +111,30 @@ private static Storage getService(String credentialsPath) throws Exception {
                 throw new RuntimeException("Failed to load Google credentials : " + credentialsPath, e);
             }
 
-            mStorageService = new Storage.Builder(httpTransport, JSON_FACTORY, credential)
+            mStorageService = new Storage.Builder(httpTransport, JSON_FACTORY, setHttpBackoffTimeout(credential))
                     .setApplicationName("com.pinterest.secor")
                     .build();
         }
         return mStorageService;
     }
 
+    private static HttpRequestInitializer setHttpBackoffTimeout(final HttpRequestInitializer requestInitializer) {
+        return new HttpRequestInitializer() {
+            @Override
+            public void initialize(HttpRequest httpRequest) throws IOException {
+                requestInitializer.initialize(httpRequest);
+
+                // Configure exponential backoff on error
+                // https://developers.google.com/api-client-library/java/google-http-java-client/backoff
+                ExponentialBackOff backoff = new ExponentialBackOff();
+                HttpUnsuccessfulResponseHandler backoffHandler = new HttpBackOffUnsuccessfulResponseHandler(backoff)
+                        .setBackOffRequired(HttpBackOffUnsuccessfulResponseHandler.BackOffRequired.ALWAYS);
+                httpRequest.setUnsuccessfulResponseHandler(backoffHandler);
+
+                httpRequest.setConnectTimeout(3 * 60000);  // 3 minutes connect timeout
```

Comment:
Maybe it's better to have those timeout value also be specifiable in config parameter, it can take 3 minutes as the default value.